usr.bin/paste/paste.c

```diff
@@ -98,51 +114,50 @@ main(int argc, char *argv[])
 		delim = tab;
 	}
 
-	if (seq)
-		rval = sequential(argv);
-	else
-		rval = parallel(argv);
+	for (filecnt = 0; *argv; ++argv, ++filecnt) {
+		failed = 0;
+		if ((lp = malloc(sizeof(LIST))) == NULL)
+			err(EXIT_FAILURE, NULL);
+		if ((*argv)[0] == '-' && !(*argv)[1])
+			lp->fp = stdin;
+		else if (!(lp->fp = fopen(*argv, "r"))) {
+			if (!seq)
+				err(EXIT_FAILURE, "%s", *argv);
+			else
+				failed = errno;
+		}
+
+		if (cap_rights_limit(fileno(lp->fp),
+		    cap_rights_init(&rights, CAP_READ)) < 0 &&
+		    errno != ENOSYS)
+			err(EXIT_FAILURE, "unable to limit rights for %s", *argv);
+
+		lp->cnt = filecnt;
+		lp->err = failed;
+		lp->name = *argv;
+
+		STAILQ_INSERT_TAIL(&lh, lp, entries);
+	}
+
+	if (cap_enter() < 0 && errno != ENOSYS)
+		err(EXIT_FAILURE, "failed to enter capability mode");
+
+	rval = seq ? sequential() : parallel();
+
```

Comment:
Do we care about freeing the list?

Reply:
It’s unnecessary to free the list explicitly since the OS will handle all memory cleanup upon program termination in this context. However, if the goal is code perfection and strict correctness, I’m open to adding a cleanup routine to free the list explicitly before exiting. But it's redundant, I believe.

Reply:
Yea, I guess its fine.

Reply:
Thanks! So I mark it as resolved.